apisix/plugins/datadog.lua

```diff
@@ -0,0 +1,277 @@
+        max_retry_count = {type = "integer", minimum = 1, default = 1},
+    }
+}
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        host = {type = "string", default= defaults.host},
+        port = {type = "integer", minimum = 0, default = defaults.port},
+        namespace = {type = "string", default = defaults.namespace},
+        constant_tags = {
+            type = "array",
+            items = {type = "string"},
+            default = defaults.constant_tags
+        }
+    },
+}
+
+local _M = {
+    version = 0.1,
+    priority = 495,
+    name = plugin_name,
+    schema = schema,
+    metadata_schema = metadata_schema,
+}
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+    return core.schema.check(schema, conf)
+end
+
+local function generate_tag(conf, const_tags)
+    local tags
+    if const_tags and #const_tags > 0 then
+        tags = core.table.clone(const_tags)
+    else
+        tags = {}
+    end
+
+    -- priority on route name, if not found using the route id.
+    if conf.route_name ~= "" then
+        core.table.insert(tags, "route_name:" .. conf.route_name)
+    elseif conf.route_id and conf.route_id ~= "" then
+        core.table.insert(tags, "route_id:" .. conf.route_id)
+    end
+
+    if conf.service_id and conf.service_id ~= "" then
+        core.table.insert(tags, "service_id:" .. conf.service_id)
+    end
+
+    if conf.consumer and conf.consumer ~= "" then
+        core.table.insert(tags, "consumer:" .. conf.consumer)
+    end
+    if conf.balancer_ip ~= "" then
+        core.table.insert(tags, "balancer_ip:" .. conf.balancer_ip)
+    end
+    if conf.response.status then
+        core.table.insert(tags, "response_status:" .. conf.response.status)
+    end
+    if conf.scheme ~= "" then
+        core.table.insert(tags, "scheme:" .. conf.scheme)
+    end
+
+    if #tags > 0 then
+        return "|#" .. concat(tags, ',')
+    end
+
+    return ""
+end
+
+-- remove stale objects from the memory after timer expires
+local function remove_stale_objects(premature)
+    if premature then
+        return
+    end
+
+    for key, batch in ipairs(buffers) do
+        if #batch.entry_buffer.entries == 0 and #batch.batch_to_process == 0 then
+            core.log.warn("removing batch processor stale object, conf: ",
+                          core.json.delay_encode(key))
+            buffers[key] = nil
+        end
+    end
+
+    stale_timer_running = false
+end
+
+function _M.log(conf, ctx)
+
+    if not stale_timer_running then
+        -- run the timer every 30 mins if any log is present
+        timer_at(1800, remove_stale_objects)
+        stale_timer_running = true
+    end
+
+    local entry = fetch_log(ngx, {})
+    entry.upstream_latency = ctx.var.upstream_response_time * 1000
+    entry.balancer_ip = ctx.balancer_ip or ""
+    entry.route_name = ctx.route_name or ""
```

Comment:
We should fall back to route_id if the name is missing. And need a test for it.

Reply:
Okay. The code block for fallback logic is at L86
Adding tests.

Reply:
The code block for fallback logic is at L86
Okay, adding tests.

Reply:
I think we can use route_name even there is only route_id? Therefore the name of tag is consistent between routes.

Reply:
Perfectly makes sense, Updated accordingly. Thanks for pointing it out.

Reply:
@spacewander ecommend using route name, won't the same routes name be confusing?

Reply:
Hi! @tzssangglass thanks for the honest concern. Yeah, in a badly designed system, it might be an issue to segregate metrics of multiple routes having the same route_name. But users can always pick multiple tags for grouping (for eg. route_name:abc, service_id:12), provided they are designing APISIX infra with service level abstraction (just one example, they can use "balancer_ip, consumer" also for filtering). 

@spacewander recommended using name over id because it's more readable to the user and there was this requirement in the Prometheus plugin (https://github.com/apache/apisix/issues/5149). Do you think that we should add a param `prefer_name` at the plugin schema or we should defer it for the upcoming versions? 
Please do let me know. Thanks.

Reply:
I get. I think it is up to the community to decide if `prefer_name` is needed like the prometheus plugin.I suggest adding a note about that  it could be misleading when route name could be duplicated for multiple.